### PR TITLE
add c code for graph operation

### PR DESCRIPTION
**Pull request by NI-SW**
If this PR is related to coding or code translation, please fill out the checklist and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

### CODE REVIEW

**Comment by NI-SW** [comment]
add c code for graph operation

File: codes/c/chapter_graph/graph_dfs.c [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    s->bot = (Vertex **)malloc(sizeof(Vertex *) * numberVertices);
+    s->top = s->bot;
+
+    return s;
+}
+
+/* 深度优先遍历 */
+void graphDFS(graphAdjList *t) {
+    // 初始化哈希表与栈
+    stack *s = newStack(t->size);
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_dfs.c:104]: 使用数据结构与其它语言不符，
- cpp 中使用 vector 存储顶点序列
- 此处使用 stack

File: codes/c/chapter_graph/graph_dfs.c [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    s->size = 0;
+    s->capacity = numberVertices;
+    s->bot = (Vertex **)malloc(sizeof(Vertex *) * numberVertices);
+    s->top = s->bot;
+
+    return s;
+}
+
+/* 深度优先遍历 */
+void graphDFS(graphAdjList *t) {
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_dfs.c:102]: 函数签名不一致
- cpp: `vector<Vertex *> graphDFS(GraphAdjList &graph, Vertex *startVet) `
- go: `func graphDFS(g *graphAdjList, startVet Vertex) []Vertex `

Comment (Gonglja) [inline codes/c/chapter_graph/graph_dfs.c:102]: 另外，将该函数拆分为两部分
- 深度优先遍历 DFS `graphDFS`
- 深度优先遍历 DFS 辅助函数 `dfs`

**Comment by Gonglja** [comment]
感谢 @NI-SW 的贡献，修改以上代码，谢谢

**Comment by NI-SW** [comment]
已修改 @Gonglja ，感谢指正。

**Comment by krahets** [comment]
Hi @NI-SW , could you plz paste the outputs of the code running to this PR?

**Comment by NI-SW** [comment]
OK~ @krahets 

gcc -g -std=c11 -o b graph_bfs.c
gcc -g -std=c11 -o d graph_dfs.c
            
-------------------------------------------------------------------BFS
 ./b.exe
初始化后，图为:
邻接表  =
0: [1, 3]
1: [0, 2, 4]
2: [1, 5]
3: [0, 4, 6]
4: [1, 3, 5, 7]
5: [2, 4, 8]
6: [3, 7]
7: [4, 6, 8]
8: [5, 7]

广度优先遍历（BFS）顶点序列为
[0, 1, 3, 2, 4, 6, 5, 7, 8]

-------------------------------------------------------------------DFS
./d.exe
邻接表  =
6: [5]
4: [5]
5: [2, 4, 6]
2: [1, 5]
3: [0]
1: [0, 2]
0: [1, 3]

深度优先遍历（DFS）顶点序列为
[0, 1, 2, 5, 4, 6, 3]

**Comment by krahets** [comment]
Thx. Pending for the check by @Gonglja

File: codes/c/chapter_graph/graph_bfs.c [hunk 3]
```diff
@@ -88,15 +88,17 @@ void freeQueue(queue *q) {
 }
 
 /* 广度优先遍历 */
-void graphBFS(graphAdjList *t) {
+int arrayIndex = 0;
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_bfs.c:91]: 该变量只在 graphBFS 函数中使用，可以将其放到内部

File: codes/c/chapter_graph/graph_bfs.c [hunk 4]
```diff
@@ -108,28 +110,26 @@ void graphBFS(graphAdjList *t) {
             }
             n = n->next;
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_bfs.c:111]: hashQuery 的返回值可以修改一下，哈希表中存在返回1，否则为 0
此处逻辑就可与 [cpp](https://github.com/krahets/hello-algo/blob/4fedf2ed705a286e19cfd0764fbfa2cdf96e3bf7/codes/cpp/chapter_graph/graph_bfs.cpp#L26-L30)逻辑一致

File: codes/c/chapter_graph/graph_dfs.c [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+/* 查询顶点是否已被标记 */
+int hashQuery(hashTable *h, int index) {
+    // 若顶点已被标记，则返回 0
+    if (h->array[index % h->size] == 1) {
+        return 0;
+    } else {
+        return 1;
+    }
+}
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_dfs.c:39]: 同上，逻辑最好反一下

File: codes/c/chapter_graph/graph_dfs.c [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+/* 释放哈希表内存 */
+void freeHash(hashTable *h) {
+    free(h->array);
+    free(h);
+}
+
+/* 深度优先遍历递归函数 */
+int arrayIndex = 0;
+void dfs(Vertex *v, hashTable *visited, Vertex **arrayVertex) {
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_dfs.c:49]: 函数签名与其它语言代码不一致，

File: codes/c/chapter_graph/graph_dfs.c [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+int main() {
+    /* 初始化无向图 */
+    graphAdjList *graph = newGraphic(10);
+    addVertex(graph, 6);
+    addVertex(graph, 4);
+    addVertex(graph, 5);
+    addVertex(graph, 2);
+    addVertex(graph, 3);
+    addVertex(graph, 1);
+    addVertex(graph, 0);
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_dfs.c:100]: 这一块没有看明白，顺序怎么是这样的？
不是 0、1、2、3、4、5、6吗

**Review by Gonglja — COMMENTED** [review]
感谢 @NI-SW 的提交，
`graph_bfs.c` 问题基本不大。
`graph_dfs.c` 初始化时存在问题

File: codes/c/chapter_graph/graph_dfs.c (hunk 7 above)
Comment (NI-SW) [reply codes/c/chapter_graph/graph_dfs.c:100]: 按顺序来会输出成这样
邻接表  =
0: [1, 3]
1: [0, 2]
2: [1, 5]
3: [0]
4: [5]
5: [2, 4, 6]
6: [5]

深度优先遍历（DFS）顶点序列为
[0, 1, 2, 5, 4, 6, 3]

File: codes/c/chapter_graph/graph_dfs.c [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+    // 释放哈希表内存并将数组索引归零
+    freeHash(visited);
+    arrayIndex = 0;
+
+    // 返回遍历数组
+    return arrayVertex;
+}
+
+/* driver code */
```
Comment (krahets) [inline codes/c/chapter_graph/graph_dfs.c:88]: /* Driver Code */

File: codes/c/chapter_graph/graph_bfs.c [hunk 9]
```diff
@@ -88,48 +88,48 @@ void freeQueue(queue *q) {
...
 }
-
+/* Driver Code */
 int main() {
 
     /* 初始化无向图 */
     graphAdjList *graph = newGraphic(3);
     // 初始化顶点
-    for (int i = 0; i < 10; i++) {
+    for (int i = 0; i < 9; i++) {
```
Comment (krahets) [inline codes/c/chapter_graph/graph_bfs.c:132]: i < 10

There are ten vertices

File: codes/c/chapter_graph/graph_dfs.c [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    // 返回遍历数组
+    return res;
+}
+
+/* Driver Code */
+int main() {
+    Vertex *initGraph[7] = {newVertex(0), newVertex(1), newVertex(2), newVertex(3),
+                            newVertex(4), newVertex(5), newVertex(6)};
+    /* 初始化无向图 */
+    graphAdjList *graph = newGraphic(10);
```
Comment (krahets) [inline codes/c/chapter_graph/graph_dfs.c:93]: Is `newGraphAdjList` more formal than `newGraphic`?

File: codes/c/chapter_graph/graph_bfs.c [hunk 11]
```diff
@@ -88,46 +88,46 @@ void freeQueue(queue *q) {
 }
 
 /* 广度优先遍历 */
-void graphBFS(graphAdjList *t) {
+Vertex **graphBFS(graphAdjList *t, Vertex *startVet) {
```
Comment (krahets) [inline codes/c/chapter_graph/graph_bfs.c:91]: The code comments are not cosistent with other languages. Please select a reference language from Java, C++ or Python and make the code and comments consistent with it.

```cpp
/* 广度优先遍历 BFS */
// 使用邻接表来表示图，以便获取指定顶点的所有邻接顶点
vector<Vertex *> graphBFS(GraphAdjList &graph, Vertex *startVet) {
    // 顶点遍历序列
    vector<Vertex *> res;
    // 哈希表，用于记录已被访问过的顶点
    unordered_set<Vertex *> visited = {startVet};
    // 队列用于实现 BFS
    queue<Vertex *> que;
    que.push(startVet);
    // 以顶点 vet 为起点，循环直至访问完所有顶点
    while (!que.empty()) {
        Vertex *vet = que.front();
        que.pop();          // 队首顶点出队
        res.push_back(vet); // 记录访问顶点
        // 遍历该顶点的所有邻接顶点
        for (auto adjVet : graph.adjList[vet]) {
            if (visited.count(adjVet))
                continue;            // 跳过已被访问过的顶点
            que.push(adjVet);        // 只入队未访问的顶点
            visited.emplace(adjVet); // 标记该顶点已被访问
        }
    }
    // 返回顶点遍历序列
    return res;
}
```

**Review by krahets — COMMENTED** [review]
Hi @NI-SW . Thanks for the PR!

I'm wondering that the consistency of your code to other languages needs to be enhanced, including:

1. Code comments.
2. Variable and function namings.
3. Algorithm process.

We make all the language consistent mainly for two reasons:

1. Easier traceability of bugs and modification.
4. Make sure the code reference in the articles can be found in ALL languages' codes.

However, I know that C is more complicated than other languages so that it more hard to keep the consistency. Let's try to make it.

File: codes/c/chapter_graph/graph_dfs.c [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+/* 释放哈希表内存 */
+void freeHash(hashTable *h) {
+    free(h->array);
+    free(h);
+}
+
+/* 深度优先遍历递归函数 */
+int resIndex = 0;
+void dfs(graphAdjList *graph, hashTable *visited, Vertex *v, Vertex **res) {
```
Comment (krahets) [inline codes/c/chapter_graph/graph_dfs.c:49]: Vertex *vet

File: codes/c/chapter_graph/graph_dfs.c [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    }
+}
+
+/* 释放哈希表内存 */
+void freeHash(hashTable *h) {
+    free(h->array);
+    free(h);
+}
+
+/* 深度优先遍历递归函数 */
```
Comment (krahets) [inline codes/c/chapter_graph/graph_dfs.c:47]: /* 深度优先遍历 DFS 辅助函数 */

File: codes/c/chapter_graph/graph_dfs.c [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+            // 递归访问邻接顶点
+            dfs(graph, visited, n->val, res);
+            n = n->next;
+        }
+    }
+    return;
+}
+
+/* 深度优先遍历 */
+Vertex **graphDFS(graphAdjList *graph, Vertex *startVet) {
```
Comment (krahets) [inline codes/c/chapter_graph/graph_dfs.c:67]: Add `// 使用邻接表来表示图，以便获取指定顶点的所有邻接顶点`

File: codes/c/chapter_graph/graph_dfs.c [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+/* 释放哈希表内存 */
+void freeHash(hashTable *h) {
+    free(h->array);
+    free(h);
+}
+
+/* 深度优先遍历递归函数 */
+int resIndex = 0;
+void dfs(graphAdjList *graph, hashTable *visited, Vertex *v, Vertex **res) {
+    // 查询哈希表，若该索引的顶点未加入数组，则加入数组并标记
```
Comment (krahets) [inline codes/c/chapter_graph/graph_dfs.c:50]: The implementation is not consistent.

```cpp
/* 深度优先遍历 DFS 辅助函数 */
void dfs(GraphAdjList &graph, unordered_set<Vertex *> &visited, vector<Vertex *> &res, Vertex *vet) {
    res.push_back(vet);   // 记录访问顶点
    visited.emplace(vet); // 标记该顶点已被访问
    // 遍历该顶点的所有邻接顶点
    for (Vertex *adjVet : graph.adjList[vet]) {
        if (visited.count(adjVet))
            continue; // 跳过已被访问过的顶点
        // 递归访问邻接顶点
        dfs(graph, visited, res, adjVet);
    }
}
```

**Comment by krahets** [comment]
Hi @NI-SW , please recheck the `graph_adjacency_list.c` and `graph_adjacency_matrix.c` in terms of consistency.

**Comment by NI-SW** [comment]
fixed @krahets  @Gonglja

**Review by krahets — APPROVED** [review]
Thanks for the PR!

**Pull request merged**